sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp

```diff
@@ -0,0 +1,928 @@
+     * @param[in] size The size of the data parameter to deserialize.
+     */
+    static AmqpValue Deserialize(uint8_t const* data, size_t size);
+
+  protected:
+    UniqueAmqpValueHandle m_value;
+  };
+  std::ostream& operator<<(std::ostream& os, AmqpValue const& value);
+
+  namespace _detail {
+
+    /** @brief Base type for AMQP collection types.
+     *
+     * Provides convenient conversions for STL collection types to enable classes derived from
+     * AmqpCollectionBase to be used as STL containers.
+     */
+    template <typename T, typename ThisType> class AmqpCollectionBase {
+    protected:
+      T m_value;
+
+      using initializer_type = std::initializer_list<typename T::value_type>;
+
+      AmqpCollectionBase(initializer_type const& initializer) : m_value{initializer} {}
+      AmqpCollectionBase(T initializer) : m_value{initializer} {}
+      AmqpCollectionBase() {}
+
+    public:
+      /** @brief Convert this collection type to an AMQP value.*/
+      operator AmqpValue() const { return static_cast<UniqueAmqpValueHandle>(*this).get(); }
+
+      /** @brief Returns the size of the underlying value.*/
+      inline typename T::size_type size() const { return m_value.size(); }
+
+      const typename T::value_type& operator[](const typename T::size_type pos) const noexcept
+      {
+        return m_value.operator[](pos);
+      }
+      void push_back(typename T::value_type&& val) { m_value.push_back(val); }
+      typename T::const_iterator begin() const noexcept { return m_value.begin(); }
+      typename T::const_iterator end() const noexcept { return m_value.end(); }
+      typename T::value_type* data() noexcept { return m_value.data(); }
+      const typename T::value_type* data() const noexcept { return m_value.data(); }
+      const typename T::value_type& at(const typename T::size_type pos) const
+      {
+        return m_value.at(pos);
+      }
+      bool operator<(ThisType const& that) const { return m_value < that.m_value; }
+      bool operator==(ThisType const& that) const { return m_value == that.m_value; }
+      bool operator!=(ThisType const& that) const { return m_value != that.m_value; }
+      /** @brief Returns true if the underlying value is empty.*/
+      bool empty() const noexcept { return m_value.empty(); }
+
+      /**
+       * @brief Convert an AmqpCollectionBase instance to a uAMQP AMQP_VALUE.
+       *
+       * @remarks This is an internal accessor and should never be used by code outside the AMQP
+       * implementation.
+       *
+       */
+      operator UniqueAmqpValueHandle() const;
+    };
+  } // namespace _detail
+
+  /** @brief Represents an AMQP array.
+   *
+   * An AMQP array is an aggregate of value types, all of which are of the same type.
+   */
+  class AmqpArray final : public _detail::AmqpCollectionBase<std::vector<AmqpValue>, AmqpArray> {
+  public:
+    /** @brief Construct a new AmqpArray object. */
+    AmqpArray() : AmqpCollectionBase(){};
```

Comment:
What would your suggestion be for how we handle a type which is not supposed to be a customer consumable type, but which is required for the public API surface?

We don't ever want customers to interact with this type directly, only through the derived types.

If this was C#, I would use the `internal` keyword, but we don't have that.

Reply:
Maybe a solution here is to use `private` inheritance instead of `public` here to hide any public surface area from the _detail base class via the public derived ones?
https://www.programiz.com/cpp-programming/public-protected-private-inheritance

Ideally, SDK authors/contributors should be able to make breaking changes to things in the `_detail` namespace without impacting end customers. If things in the `_detail` namespace end up being base classes for public types, any public surface area (like a virtual/pure virtual function) can no longer be renamed/changed/modified without breaking an end-customer who could be inadvertently calling it.

Reply:
See related issue about coming up with guidance whether we allow public types to inherit from types in the `_detail` namespace (answer - no, with DateTime being an exception):
https://github.com/Azure/azure-sdk-for-cpp/issues/1599#issuecomment-858097658